tracee-ebpf/tracee/process_tree.go

```diff
@@ -0,0 +1,174 @@
+package tracee
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/aquasecurity/tracee/pkg/containers"
+)
+
+type ProcessCtx struct {
+	StartTime   int
+	ContainerID string
+	Pid         uint32
+	Tid         uint32
+	Ppid        uint32
+	HostTid     uint32
+	HostPid     uint32
+	HostPpid    uint32
+	Uid         uint32
+	MntId       uint32
+	PidId       uint32
+}
+
+type ProcessTree struct {
+	processTreeMap map[int]ProcessCtx
+}
+
+func (t *Tracee) ParseProcessContext(ctx []byte) (ProcessCtx, error) {
+	var procCtx = ProcessCtx{}
+	procCtx.StartTime = int(binary.LittleEndian.Uint64(ctx[0:8]))
+	cgroupId := binary.LittleEndian.Uint64(ctx[8:16])
+	procCtx.ContainerID = t.containers.GetCgroupInfo(cgroupId).ContainerId
+	dataBuff := bytes.NewBuffer(ctx[16:]) // this is the offset after the cgroup and startTime in the ctx byte array
+	var errs []error
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.Pid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.Tid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.Ppid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.HostTid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.HostPid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.HostPpid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.Uid))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.MntId))
+	errs = append(errs, binary.Read(dataBuff, binary.LittleEndian, &procCtx.PidId))
+	for _, e := range errs {
+		if e != nil {
+			return procCtx, e
+		}
+	}
+	return procCtx, nil
+}
+
+//returns the file creation time
+//Note: this is not the real process start time because it's based on the file creation time, but it is a close approximation
+func getFileCtime(path string) (int, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return 0, err
+	}
+	stat_t := fileInfo.Sys().(*syscall.Stat_t)
+	ts := stat_t.Ctim
+	return int(ts.Sec)*int(time.Second) + int(ts.Nsec)*int(time.Nanosecond), nil
+}
+
+//parse the status file of a process or given task and returns process context struct
+func parseProcStatus(status []byte, taskStatusPath string) (ProcessCtx, error) {
+	var process ProcessCtx
+	processVals := make(map[string]uint32)
+	for _, line := range strings.Split(string(status), "\n") {
+		lineFields := strings.Fields(line)
```

Comment:
Why not using `lineFields := strings.Split(line, ":")` to have the keys without a colon?